src/core/mesh/qgsmeshdataprovidertemporalcapabilities.h

```diff
@@ -0,0 +1,139 @@
+#ifndef QGSMESHDATAPROVIDERTEMPORALCAPABILITIES_H
+#define QGSMESHDATAPROVIDERTEMPORALCAPABILITIES_H
+
+#include "qgsdataprovidertemporalcapabilities.h"
+#include "qgsrange.h"
+#include "qgsmeshdataset.h"
+
+
+/**
+ * \class QgsMeshDataProviderTemporalCapabilities
+ * \ingroup core
+ * Class for handling properties relating to a mesh data provider's temporal capabilities.
+ *
+ * \since QGIS 3.14
+ */
+class CORE_EXPORT QgsMeshDataProviderTemporalCapabilities: public QgsDataProviderTemporalCapabilities
+{
+  public:
+    /**
+     * Constructor for QgsMeshDataProviderTemporalCapabilities
+     */
+    QgsMeshDataProviderTemporalCapabilities();
+
+    /**
+     * Returns the last dataset index that has the time equal or lesser or than \a timeSinceGlobalReference (in milliseconds)from the dataset \a group
+     *
+     * Returns invalide dataset index if the time is less than the first dataset's time
+     */
+    QgsMeshDatasetIndex datasetIndexTimeInMilliseconds( int group, qint64 timeSinceGlobalReference ) const;
+
+    /**
+     * Returns the last dataset index that has the time equal or lesser or than \a timeSinceGlobalReference (in hours)from the dataset \a group
+     *
+     * Returns invalide dataset index if the time is less than the first dataset's time
+     */
+    QgsMeshDatasetIndex datasetIndex( int group, double  timeSinceGlobalReferenceInHours ) const;
+
+    /**
+     * Adds a \a reference date/time from a dataset \a group
+     *
+     * \note must be used only by the mesh data provider
+     */
+    void addGroupReferenceDateTime( int group, const QDateTime &reference ) SIP_SKIP;
+
+    /**
+     * Adds a \a time (in miliseonds) from a dataset contained in \a group
+     *
+     * \note must be used only by the mesh data provider,
+     * all dataset need to be added one after one
+     */
+    void addDatasetTimeInMilliseconds( int group, qint64 time ) SIP_SKIP;
+
+    /**
+     * Adds a \a time (in provider unit) from a dataset contained in \a group
+     *
+     * \note must be used only by the mesh data provider,
+     * all dataset need to be added one after one
+     */
+    void addDatasetTime( int group, double  time ) SIP_SKIP;
+
+    /**
+     * Returns true if the data has a reference time
+     *
+     */
+    bool hasReferenceTime() const;
+
+    /**
+     * Returns the reference time
+     *
+     */
+    QDateTime referenceTime() const;
+
+    /**
+     * Returns the time extent using the internal reference time
+     * and the first and last times available from the all the dataset
+     *
+     */
+    QgsDateTimeRange timeExtent() const;
+
+    /**
+     * Returns the time extent using \a reference date time
+     * and the first and last times available from the all the dataset
+     */
+    QgsDateTimeRange timeExtent( const QDateTime &reference ) const;
```

Comment:
For reference time I would create setter (setReferenceTime) and remove this functions. Then `QgsDateTimeRange timeExtent() const;` will return absolute time (and if dataset is relative automatically use reference time)

Reply:
The principle is here that the provider capabilities has a reference time only if the data provides a reference time. So we couldn't not permit to change this reference time by other part than the data provider.
The reference time coming from user of from project is store in the temporal properties of the layer.
So the `timeExtent(const QDateTime &reference )` is used when the time extent is needed with an external reference time, and `timeExtent()` is used to have the "raw" extent time of the data if data reference time exists